x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -175,6 +177,45 @@ protected static List<Batch<LogicalPlan>> rules() {
         return asList(substitutions, operators(), skip, cleanup(), defaultTopN, label);
     }
 
+    static class InvertSpatialFunctions extends OptimizerRules.OptimizerRule<UnaryPlan> {
+
+        InvertSpatialFunctions() {
+            super(TransformDirection.UP);
+        }
+
+        @Override
+        protected LogicalPlan rule(UnaryPlan plan) {
+            if (plan instanceof Filter filter) {
+                var condition = filter.condition().transformDown(SpatialRelatesFunction.class, this::invertIfNecessary);
+                if (filter.condition().equals(condition) == false) {
+                    plan = new Filter(filter.source(), filter.child(), condition);
+                }
+            }
+            if (plan instanceof Eval eval) {
+                List<Alias> fields = eval.fields();
+                List<Alias> changed = fields.stream()
+                    .map(f -> (Alias) f.transformDown(SpatialRelatesFunction.class, this::invertIfNecessary))
+                    .toList();
+                if (changed.equals(fields) == false) {
+                    plan = new Eval(eval.source(), eval.child(), changed);
+                }
+            }
+            return plan;
+        }
+
+        private SpatialRelatesFunction invertIfNecessary(SpatialRelatesFunction srf) {
+            if (srf.isCommutative()) {
+                // Never bother to invert commutative functions
+                return srf;
+            }
+            if (srf.left().foldable() && srf.right().foldable() == false) {
+                // If we have a constant on the left, and a field on the right, invert the function
```

Comment:
1. Why is this necessary/helpful?
2. Is this really equivalent in all cases? It looks to me like inverting e.g. `SpatialWithin` or `SpatialContains` produces a non-equivalent condition. Don't we need to negate the whole thing for this to work?

Reply:
Scratch No.2 - I just realized that within/contains gets replaced by the other, resp.